sky/server/daemons.py

```diff
@@ -126,14 +135,53 @@ def refresh_volume_status_event():
     time.sleep(server_constants.VOLUME_REFRESH_DAEMON_INTERVAL_SECONDS)
 
 
+_managed_job_consolidation_mode_lock = None
+
+
 def managed_job_status_refresh_event():
     """Refresh the managed job status for controller consolidation mode."""
     # pylint: disable=import-outside-toplevel
+    from sky.jobs import constants as managed_job_constants
     from sky.jobs import utils as managed_job_utils
 
-    # We run the recovery logic before starting the event loop as those two are
-    # conflicting. Check PERSISTENT_RUN_RESTARTING_SIGNAL_FILE for details.
-    managed_job_utils.ha_recovery_for_consolidation_mode()
+    global _managed_job_consolidation_mode_lock
+    if _managed_job_consolidation_mode_lock is None:
+        _managed_job_consolidation_mode_lock = locks.get_lock(
+            managed_job_constants.CONSOLIDATION_MODE_LOCK_ID)
+
+    # Touch the signal file here to avoid conflict with
+    # update_managed_jobs_statuses. Although we run
+    # ha_recovery_for_consolidation_mode before checking the job statuses
+    # (events.ManagedJobEvent), update_managed_jobs_statuses is also called in
+    # cancel_jobs_by_id.
+    # We also need to make sure that new controllers are not started until we
+    # acquire the consolidation mode lock, since if we have controllers on both
+    # the new and old API server during a rolling update, calling
+    # update_managed_jobs_statuses on the old API server could lead to
+    # FAILED_CONTROLLER.
+    signal_file = pathlib.Path(
+        constants.PERSISTENT_RUN_RESTARTING_SIGNAL_FILE).expanduser()
+    try:
+        signal_file.touch()
+
+        # Make sure the lock is acquired for this process before proceeding to
+        # do recovery. This will block if another API server is still running,
+        # but should proceed once it is terminated and releases the lock.
+        if not _managed_job_consolidation_mode_lock.is_locked():
+            logger.info('Acquiring the consolidation mode lock: '
+                        f'{_managed_job_consolidation_mode_lock}')
+            _managed_job_consolidation_mode_lock.acquire()
```

Comment:
Is this a distribueted lock? not exactly sure when will this be released.

Curious, considering we may have multiple replica of API server in the future, should we just do a standard leader election, e.g. use some lease in our managed job table (or even a global table for a leader API server)?